OpenRA.Mods.Common/Activities/DeployForGrantedCondition.cs

```diff
@@ -18,44 +18,60 @@ namespace OpenRA.Mods.Common.Activities
 	public class DeployForGrantedCondition : Activity
 	{
 		readonly GrantConditionOnDeploy deploy;
+		readonly IMove move;
 		readonly bool canTurn;
+		readonly bool orderedMove;
+		readonly CPos cell;
+		bool initiated;
 
-		public DeployForGrantedCondition(Actor self, GrantConditionOnDeploy deploy)
+		public DeployForGrantedCondition(Actor self, GrantConditionOnDeploy deploy, Target target)
 		{
 			this.deploy = deploy;
 			canTurn = self.Info.HasTraitInfo<IFacingInfo>();
+			orderedMove = target.Type == TargetType.Terrain || (target.Type == TargetType.Actor && target.Actor != self);
+			if (orderedMove)
+				cell = self.World.Map.Clamp(self.World.Map.CellContaining(target.CenterPosition));
+
+			move = self.TraitOrDefault<IMove>();
 		}
 
 		protected override void OnFirstRun(Actor self)
 		{
 			// Turn to the required facing.
-			if (deploy.Info.Facing != -1 && canTurn)
+			if (deploy.DeployState == DeployState.Undeployed && deploy.Info.Facing != -1 && canTurn && !orderedMove)
 				QueueChild(self, new Turn(self, deploy.Info.Facing));
 		}
 
 		public override Activity Tick(Actor self)
 		{
-			// Do turn first, if needed.
 			if (ChildActivity != null)
 			{
 				ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
 				return this;
 			}
 
-			// Without this, turn for facing deploy angle will be canceled and immediately deploy!
 			if (IsCanceling)
 				return NextActivity;
 
-			if (IsInterruptible)
+			if (!initiated)
 			{
-				IsInterruptible = false; // must DEPLOY from now.
-				deploy.Deploy();
-				return this;
-			}
+				initiated = true;
+				if (deploy.DeployState == DeployState.Undeployed)
+				{
+					if (orderedMove && move != null)
+						QueueChild(self, move.MoveTo(cell, 8), true);
```

Comment:
Hmm, do I get this right? When we're taking this branch, no deploy will ever actually be executed, correct? So this is basically just wrapping around a regular move order for undeployed units? Is that intentional?

Reply:
Yes, this is intentional. All move orders are intercepted by `GrantConditionOnDeploy` and interpreted on the fly by `DeployForGrantedCondition` as either regular moves or undeploy-moves depending on the state at the moment the activity is executed. This ensures that move orders are always interpreted correctly no matter how convoluted the activity-queue gets.